crates/bevy_reflect/src/path/access.rs

```diff
@@ -0,0 +1,240 @@
+            ReflectRef::Map(_) => Type::Map,
+            ReflectRef::Enum(_) => Type::Enum,
+            ReflectRef::Value(_) => Type::Value,
+        }
+    }
+}
+impl From<VariantType> for Type {
+    fn from(value: VariantType) -> Self {
+        match value {
+            VariantType::Struct => Type::Struct,
+            VariantType::Tuple => Type::Tuple,
+            VariantType::Unit => Type::Unit,
+        }
+    }
+}
+
+/// A singular owned element access within a path.
+///
+/// Can be applied to a `dyn Reflect` to get a reference to the targeted element.
+///
+/// A path is composed of multiple accesses in sequence.
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub(super) enum Access {
+    Field(Box<str>),
+    FieldIndex(usize),
+    TupleIndex(usize),
+    ListIndex(usize),
+}
+
+impl fmt::Display for Access {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Access::Field(field) => write!(f, ".{field}"),
+            Access::FieldIndex(index) => write!(f, "#{index}"),
+            Access::TupleIndex(index) => write!(f, ".{index}"),
+            Access::ListIndex(index) => write!(f, "[{index}]"),
+        }
+    }
+}
+
+/// A singular borrowed element access within a path.
+///
+/// Can be applied to a `dyn Reflect` to get a reference to the targeted element.
+///
+/// Does not own the backing store it's sourced from.
+/// For an owned version, you can convert one to an [`Access`] with [`AccessRef::to_owned`].
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(super) enum AccessRef<'a> {
+    Field(&'a str),
+    FieldIndex(usize),
+    TupleIndex(usize),
+    ListIndex(usize),
+}
+
+impl Access {
+    pub(super) fn as_ref(&self) -> AccessRef<'_> {
+        match self {
+            Self::Field(value) => AccessRef::Field(value),
+            Self::FieldIndex(value) => AccessRef::FieldIndex(*value),
+            Self::TupleIndex(value) => AccessRef::TupleIndex(*value),
+            Self::ListIndex(value) => AccessRef::ListIndex(*value),
+        }
+    }
+}
+
+impl<'a> AccessRef<'a> {
+    pub(super) fn to_owned(self) -> Access {
+        match self {
+            Self::Field(value) => Access::Field(value.to_string().into_boxed_str()),
+            Self::FieldIndex(value) => Access::FieldIndex(value),
+            Self::TupleIndex(value) => Access::TupleIndex(value),
+            Self::ListIndex(value) => Access::ListIndex(value),
+        }
+    }
+
+    fn display_value(&self) -> &dyn fmt::Display {
+        match self {
+            Self::Field(value) => value,
+            Self::FieldIndex(value) | Self::TupleIndex(value) | Self::ListIndex(value) => value,
+        }
+    }
+    fn kind(&self) -> &'static str {
+        match self {
+            Self::Field(_) => "field",
+            Self::FieldIndex(_) => "field index",
+            Self::TupleIndex(_) | Self::ListIndex(_) => "index",
+        }
+    }
+
+    pub(super) fn element(
```

Comment:
Nit: Should we call these methods `into_element` since it takes `self`? Not a big deal if we don't, but the `into_` prefix makes it clear at a glance that we're consuming the type.

Reply:
The important bit here is that `AccessRef` is `Copy`, this is why `element` takes self. Due to the copy nature of `AccessRef`, it's not consuming, so `into_` is a bit misleading.